pkg/rbd/nodeserver.go

```diff
@@ -40,59 +40,91 @@ type nodeServer struct {
 
 func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublishVolumeRequest) (*csi.NodePublishVolumeResponse, error) {
 	targetPath := req.GetTargetPath()
-
-	if !strings.HasSuffix(targetPath, "/mount") {
-		return nil, fmt.Errorf("rnd: malformed the value of target path: %s", targetPath)
-	}
-	s := strings.Split(strings.TrimSuffix(targetPath, "/mount"), "/")
-	volName := s[len(s)-1]
-
 	targetPathMutex.LockKey(targetPath)
 	defer targetPathMutex.UnlockKey(targetPath)
 
-	notMnt, err := ns.mounter.IsLikelyNotMountPoint(targetPath)
-	if err != nil {
-		if os.IsNotExist(err) {
-			if err = os.MkdirAll(targetPath, 0750); err != nil {
-				return nil, status.Error(codes.Internal, err.Error())
+	var volName string
+	isBlock := req.GetVolumeCapability().GetBlock() != nil
+
+	if isBlock {
+		// Get volName from targetPath
+		s := strings.Split(targetPath, "/")
+		volName = s[len(s)-1]
+
+		// Check if that target path exists properly
+		// targetPath should exists and should be a file
+		st, err := os.Stat(targetPath)
+		if err != nil {
+			if os.IsNotExist(err) {
+				return nil, status.Error(codes.NotFound, "targetPath not mounted")
 			}
-			notMnt = true
-		} else {
 			return nil, status.Error(codes.Internal, err.Error())
 		}
-	}
+		if !st.Mode().IsRegular() {
+			return nil, status.Error(codes.Internal, "targetPath is not regular file")
```

Comment:
could the target path be a sym link? if so, should the sym link be resolved first and check if it is a regular file?

Reply:
It should not be a symlink and it should be a regular file that CO create just for bind mount the device. So, keeping it as is.